server/src/daemon/service.rs

```diff
@@ -0,0 +1,391 @@
+//  org.freedesktop.Secret.Service
+
+use std::{
+    collections::HashMap,
+    env, fs,
+    sync::{Arc, Mutex},
+    time::SystemTime,
+};
+
+use oo7::{
+    dbus::{
+        api::{Properties, SecretInner},
+        Algorithm,
+    },
+    portal::{Item, Keyring, Secret},
+    Key,
+};
+use tokio::sync::RwLock;
+use zbus::{
+    proxy::ProxyDefault,
+    zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Value},
+    ObjectServer, SignalContext,
+};
+
+use super::{
+    collection::Collection, error::ServiceError, item, prompt::Prompt,
+    service_manager::ServiceManager, session::Session, Result,
+};
+#[cfg(debug_assertions)]
+use crate::SERVICE_NAME;
+use crate::{LOGIN_KEYRING, LOGIN_KEYRING_PATH};
+
+#[derive(Clone, Debug)]
+pub struct Service {
+    collections: Arc<RwLock<Vec<Collection>>>,
+    keyring: Arc<Keyring>,
+    cnx: Arc<Mutex<Option<zbus::Connection>>>,
+    manager: Arc<Mutex<ServiceManager>>,
+    sessions_counter: Arc<RwLock<i32>>,
+}
+
+#[zbus::interface(name = "org.freedesktop.Secret.Service")]
+impl Service {
+    pub async fn open_session(
+        &self,
+        algorithm: Algorithm,
+        input: Value<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(OwnedValue, OwnedObjectPath)> {
+        let (service_public_key, aes_key) = match algorithm {
+            Algorithm::Plain => (None, None),
+            Algorithm::Encrypted => {
+                let private_key = Key::generate_private_key();
+                let client_public_key = Key::from(input);
+                (
+                    Some(Key::generate_public_key(&private_key)),
+                    Some(Key::generate_aes_key(&private_key, &client_public_key)),
+                )
+            }
+        };
+
+        *self.sessions_counter.write().await += 1;
+        let session = Session::new(
+            aes_key,
+            Arc::clone(&self.manager),
+            *self.sessions_counter.read().await,
+        );
+        self.manager
+            .lock()
+            .unwrap()
+            .insert_session(session.path().to_owned(), session.to_owned());
+
+        object_server
+            .at(session.path().to_owned(), session.to_owned())
+            .await?;
```

Comment:
If you create the path here, you would avoid a bit of dancing here with the path.